Update geobounds aggregation. (#9426)
opensearch-project/documentation-website: _aggregations/metric/geobounds.md

  - /query-dsl/aggregations/metric/geobounds/
---

# Geobounds aggregations

> The `geo_bounds` metric is a multi-value metric that calculates the [geographic bounding box](https://docs.ogc.org/is/12-063r5/12-063r5.html#30) encompassing all of a set of `geo_point` or `geo_shape` objects. The bounding box is returned as the upper-left and lower-right vertices of the rectangle given as a decimal-encoded latitude-longitude (lat-lon) pair.

Reviewer: Could you link to geopoint and geoshape pages? https://opensearch.org/docs/latest/field-types/supported-field-types/geo-point/
https://opensearch.org/docs/latest/field-types/supported-field-types/geo-shape/

> The `geo_bounds` metric is a multi-value metric that calculates the [geographic bounding box](https://docs.ogc.org/is/12-063r5/12-063r5.html#30) encompassing all of a set of `geo_point` or `geo_shape` objects. The bounding box is returned as the upper-left and lower-right vertices of the rectangle given as a decimal-encoded latitude-longitude (lat-lon) pair.

Author: ```suggestion
The `geo_bounds` aggregation is a multi-value aggregation that calculates the [geographic bounding box](https://docs.ogc.org/is/12-063r5/12-063r5.html#30) encompassing a set of [`geo_point`](https://opensearch.org/docs/latest/field-types/supported-field-types/geo-point/) or [`geo_shape`](https://opensearch.org/docs/latest/field-types/supported-field-types/geo-shape/) objects. The bounding box is returned as the upper-left and lower-right vertices of the rectangle given as a decimal-encoded latitude-longitude (lat-lon) pair.
```

## Parameters

> The `geo_bounds` aggregation takes the following parameters: 

Reviewer: ```suggestion
The `geo_bounds` aggregation takes the following parameters.
```

| Parameter        | Required/Optional | Data type      | Description |
| :--              | :--               | :--            | :--         |

> | `field`          | Required          | String         | Name of the field containing the geopoints or geoshape for which the geobound is computed. |

Reviewer: ```suggestion
| `field`          | Required          | String         | The name of the field containing the geopoints or geoshapes for which the geobounds are computed. |
```

Author: It's not clear to me whether you can provide more than one `geoshape` to the aggregation?

Reviewer: @rishabhmaurya Is it possible to provide more than one geoshape to the aggregation?

Reviewer 2: that's the name of the field and its accepting name of the field on which one is aggregating

Reviewer: Thanks! @dwelsch-esi For every document, there is one field that contains either a geopoint or a geoshape. So I think using plural for both geopoint and geoshape is fine because it conveys an aggregation on a set of geopoints and geoshapes for all documents.

> | `wrap_longitude` | Optional          | Boolean        | Whether to allow the bounding box to overlap the international date line. The default is true. |

Reviewer: ```suggestion
| `wrap_longitude` | Optional          | Boolean        | Whether to allow the bounding box to overlap the international date line. Default is `true`. |
```

## Example

> This example returns the `geo_bounds` for the `geoip.location` of every order in the ecommerce sample data (each `geoip.location` is a geopoint):

Reviewer: ```suggestion
This example returns the `geo_bounds` for the `geoip.location` of every order in the e-commerce sample data (each `geoip.location` is a geopoint):
```

  }
}
```
{% include copy-curl.html %}

## Example response

As shown in the following example response, the aggregation returns the `geobounds` containing all geopoints in the `geoip.location` field:

```json
{
  "took": 16,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 4675,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  },
  "aggregations": {
    "geo": {
      "bounds": {
        "top_left": {
          "lat": 52.49999997206032,
          "lon": -118.20000001229346
        },
        "bottom_right": {
          "lat": 4.599999985657632,
          "lon": 55.299999956041574
        }
      }
    }
  }
}
```

## Aggregating geoshapes

You can run a `geo_bounds` aggregation on geoshapes.

> Prepare an example by inserting an index containing a geoshape field:

Reviewer 3: @dwelsch-esi Got it. Makes sense to me 👍

> Prepare an example by inserting an index containing a geoshape field:

Author: ```suggestion
Prepare an example by inserting an index containing a geoshape field. GeoJSON input specifies longitude first:
```

  }
}
```
{% include copy-curl.html %}

## Example response

> The response is as follows:

Reviewer: Could you provide a descriptive sentence for the response?

Author: ```suggestion
As shown in the following example response, the aggregation returns the `geobounds` containing all `geopoints` in the `geoip.location` field:
```

```json
{
  "took": 16,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 4675,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  },
  "aggregations": {
    "geo": {
      "bounds": {
        "top_left": {
          "lat": 52.49999997206032,
          "lon": -118.20000001229346
        },
        "bottom_right": {
          "lat": 4.599999985657632,
          "lon": 55.299999956041574
        }
      }
    }
  }
}
```

## Aggregating geoshapes

You can run a `geo_bounds` aggregation on geoshapes.

> Prepare an example by inserting an index containing a geoshape field:

Reviewer: Please split the example into 2 code blocks:

1. Prepare an example by creating an index containing a geoshape field
2. Ingest documents into the index
```

Author: ```suggestion
Prepare an example by inserting an index containing a geoshape field. GeoJSON input specifies longitude first:
```

> GeoJSON input specifies longitude first.

Reviewer: Could you move this note so the intro sentence is immediately before the code block. Alternatively, you can just create a 2-sentence intro and end this one with a colon.

> 
> GeoJSON input specifies longitude first.

Author: ```suggestion
```

> 
> GeoJSON input specifies longitude first.
> {: .note}

Author: ```suggestion
```

    }
  }
}

> 

Reviewer: ````suggestion
```
{% include copy-curl.html %}
````

> POST _bulk

Reviewer: ````suggestion

```json
POST _bulk
````

Author: ```suggestion
Ingest documents into the index:

```json
POST _bulk
```

```json
{
  "took": 8,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 4,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  },
  "aggregations": {
    "grouped": {
      "bounds": {
        "top_left": {
          "lat": 45.11999997776002,
          "lon": 144.71999991685152
        },
        "bottom_right": {
          "lat": 13.469999986700714,
          "lon": -109.83000006526709
        }
      }
    }
  }
}
```

## Wrapping longitude

If the optional `wrap_longitude` parameter is set to `true`, the bounding box can overlap the international date line (180&deg; meridian) and return a `bounds` object in which the top-left longitude is greater than the bottom-right longitude. The default value for `wrap_longitude` is `true`.

Re-run the `geo_bounds` aggregation on the national parks geoshape with `wrap_longitude` set to `false`:

```json
GET national_parks/_search
{
  "size": 0,
  "aggregations": {
    "grouped": {
      "geo_bounds": {
        "field": "location",
        "wrap_longitude": false
      }
    }
  }
}
```
{% include copy-curl.html %}

> Note that the new resulting geobound encompasses a larger area to avoid overlapping the dateline.

Reviewer: ```suggestion
Note that the new resulting geobound encompasses a larger area to avoid overlapping the dateline:
```

```
{% include copy-curl.html %}

> Ingest documents into the index:

Author: ```suggestion
Ingest documents into the index. GeoJSON input specifies longitude first:
```

```json
POST _bulk
{ "create": { "_index": "national_parks", "_id": "1" } }
{"name": "Yellowstone National Park", "location": {"type": "envelope","coordinates": [ [-111.15, 45.12], [-109.83, 44.12] ]}}
{ "create": { "_index": "national_parks", "_id": "2" } }
{ "name": "Yosemite National Park", "location": {"type": "envelope","coordinates": [ [-120.23, 38.16], [-119.05, 37.45] ]} }
{ "create": { "_index": "national_parks", "_id": "3" } }
{ "name": "Death Valley National Park", "location": {"type": "envelope","coordinates": [ [-117.34, 37.01], [-116.38, 36.25] ]} }
{ "create": { "_index": "national_parks", "_id": "4" } }
{ "name": "War In The Pacific National Historic Park Guam", "location": {"type": "point","coordinates": [144.72, 13.47]} }
```
{% include copy-curl.html %}

Run a `geo_bounds` aggregation on the `location` field:

```json
GET national_parks/_search
{
  "size": 0,
  "aggregations": {
    "grouped": {
      "geo_bounds": {
        "field": "location",
        "wrap_longitude": true
      }
    }
  }
}
```
{% include copy-curl.html %}

> The response contains the smallest geo-bounding box that encloses all shapes in the `location` field:

Reviewer 3: @dwelsch-esi I agree with your logic, and given that we use "geo-bounding" elsewhere, I think this is fine as is.
